src/base/Optional.h

```diff
@@ -60,10 +73,33 @@ class Optional
         return *this;
     }
 
+    void clear() {
+        if (hasValue_) {
+            value_.~Value();
+            hasValue_ = false;
+        }
+    }
+
 private:
-    Value value_; // stored value; inaccessible/uninitialized unless hasValue_
+    union {
+        unsigned char dummy_;
```

Comment:
I assume that this dummy_ field is needed so that we can initialize it inside the union in the default Optional constructor:

* Without this field and without value_ initialization, the compiler will complain that the union is left uninitialized.
* Without this field but with value_ initialization, we will create a potentially non-trivial value_ object which may be expensive and goes against Optional "the value does not exist" default.
* We cannot use in-class `dummy_ = 0` initialization because the compiler will complain that we are initializing both union members in non-default Optional constructor and/or because this initialization would be needlessly initializing value_ memory in non-default Optional constructor (that also initializes value_ memory).
* The dummy_ type  can be `char` or `byte8_t` or any sizeof=1 type. There is no special significance in using _unsigned_ char.

Please correct any wrong assumptions.

Reply:
> Without this field and without value_ initialization, the compiler will complain 

This limitation is forced by constexpr constructor requirements, as outlined [here](https://stackoverflow.com/questions/57496628/why-is-there-a-dummy-union-member-in-some-implementations-of-
stdoptional)

> Without this field but with value_ initialization, we will create a potentially non-trivial value_ object which may be expensive

Agreed.

> We cannot use in-class dummy_ = 0 initialization because the compiler will complain that we are initializing both union members in non-default

My test showed that the compiler was OK with that.

> this initialization would be needlessly initializing value_ memory in non-default Optional constructor

I think it is this case that is problematic.

> The dummy_ type can be char or byte8_t or any sizeof=1 type. There is no special significance in using unsigned char.

Agreed, we can use `bool` as well, I think.

Reply:
I extracted the Optional enhancement from this branch into a dedicated branch and posted official PR 1106. Unfortunately, I had to fix a few bugs, so there is an elevated probability that I screwed something up in the process. Please check that the Optional in the official PR works for this internal PR and double check the following bug-fixing commits on that dedicated branch:

* 881a830 fixup: Fixed Optional move assignment
* 0c5d8cf fixup: Fix incorrectly defaulted copy constructor and copy assignment

These commits may also deserve a second look (but they do not fix any bugs):

* 84de3ca fixup: No need to touch dummy_ (except when default-initializing)
* 2759cde fixup: Simplified dummy_ type

Reply:
> Unfortunately, I had to fix a few bugs,

I am not sure that these two were bugs. In this issue, IIRC, we did not intend  to develop a general-purpose STL-compilant Optional class, but rather improve it a little to support RandomUuid, with its limitations. For example, RandomUuid does not suport a default/unitialized state - it gets always a meanigful ID. That is why Optional move operator had to clear() - for otherwise the moved from Optional would have no indication of 'no value'. As for copy constructors - since RandomUuid does not support copy semantics, that functionality was intentionally omitted (until it is required somewhere else).
Probably we should have developed a standard-compliant Optional (for non-trivial types) first - as a stand-alone issue - and _then_ see whether it could be applied for RandomUuid. But again, we focused here on the issue specifics, rather than general-purpose tools.

Reply:
> In this issue, IIRC, we did not intend to develop a general-purpose STL-compilant Optional class, but rather improve it a little to support RandomUuid, 

Correct.

> with its limitations.

Not sure what you mean by that exactly, but the result should (at the very least) correctly support anything that compiles. Anything short of that is a bug. While RandomUuid was the trigger for these enhancements, the resulting Optional implementation would, of course, (and has already been!) used for other non-trivial classes (e.g., Ip::Address).

It may be possible to wiggle out of the bug label the "incorrectly defaulted copy constructor and copy assignment" problem because that problem led to compile errors[^1], but the extra reset() problem fixed in 0c5d8cfad4e0b45478732559792f46892e0f478e was a bug because it affects Optional semantics _regardless_ of the Value type -- we broke that semantics even for trivial types!

[^1]: I do not know whether all other use cases where the generated implementation would be wrong would result in a compile error, but let's make that generous assumption for the sake of the argument.

> For example, RandomUuid does not suport a default/unitialized state - it gets always a meanigful ID. That is why Optional move operator had to clear() - for otherwise the moved from Optional would have no indication of 'no value'. 

What you are saying is logical, but C++ move and std::optional use different logic:

* The post-move state of a RandomUuid object is _irrelevant_ because it is essentially undefined for any RandomUuid-using code. Our code must not use the RandomObject after the move (except for destroying it or resetting it to something else) -- any such use is undefined behavior! (Usually, there is no such use _naturally_ because the variable goes out of scope when moved, but an _explicit_ std::move() call increases UB risks.)
* The post-move state of a RandomUuid object that you are talking about is _not_ the responsibility of Optional code. It is the responsibility of RandomUuid code! Optional simply uses the move facilities provided by its Value class. All your "meaningful ID" arguments (if we assume, for the sake of the argument, that they are correct) can only apply to RandomUuid code, not Optional code.
* Please also note that the code was "clearing" not just the moved RamdomUuid object that you are talking about above but its Optional object _owner_, directly violating std::optional semantics. Yes, that semantics may feel odd, but we have to mimic it -- a valued Optional must remain non-empty after the move.

IIRC, there was another related bug: The new public method name -- clear() -- did not match std::optional::reset() name. Some day, we want to remove Optional. Besides the class name change and unavoidable namespace transition side-effects, the then-existing Squid code should continue to work "as is". We do not want to go through the code and rename clear() to reset() at that time.

Reply:
> Fix incorrectly defaulted copy constructor and copy assignment

In the implemented assignment operator= we now call the existing move operator=:  `*this = other.value_`. Looks a little confusing, but probably it should work as intended (because the move operator uses std::forward() and the lvalue should be copied in this case).

Reply:
```C++
    Optional &operator =(const Optional &other)
    {
        if (this != &other) {
            if (other.hasValue_)
                *this = other.value_;
```

> In the implemented assignment operator= we now call the existing move operator=: `*this = other.value_`

Yes, I hope so (although we probably should not call that operator a _move_ operator because, if we are right about how universal references work, that operator does not necessarily _move_ the otherValue):

```
    template <class Other = Value>
    Optional &operator =(Other &&otherValue)
    {
        value_ = std::forward<Other>(otherValue); // this is either a value copy or a value move
        hasValue_ = true;
        return *this;
    }
```

> Looks a little confusing

I very much agree! I could not decide whether to wrap all these low-level copy/move operations into _named_ methods, so that operators and constructors can call those methods by name instead of betting that the right overloaded method will be called. I am still undecided.

> probably it should work as intended (because the move operator uses std::forward() and the lvalue should be copied in this case

Yes, I hope that code works that way, but I do not like the fact that I have to think about this and am not sure my conclusion is correct. I do not know whether I am not sure because I am just not good enough with universal references yet OR because the code (is poorly written and) can be improved (e.g., by introducing private named methods instead of relying exclusively on the overloaded assignment operators).

I am leaning towards the former, especially when I read the subtitle in the well-known Meyers [talk](https://isocpp.org/blog/2012/11/universal-references-in-c11-scott-meyers) about universal references: T&& Doesn’t Always Mean “Rvalue Reference”. We should get used to the universal reference concept, so that we do not, in Meyers' words, "misread a lot of C++11 code"! I am not quite there yet, but this last exchange will help me. Thank you.